### ISSUE TITLE
[Hooks] An official global state-management solution (a la `redux`)

### ISSUE THREAD

**Issue by derrickbeining** [reactjs/rfcs#90]
Related to #68 
I'm unsure if this is the appropriate place for this. If not, feel free to close.

The `useState` hook feels very similar to the way [atom](https://purelyfunctional.tv/guide/reagent/#atoms)s work in [reagent.cljs](https://reagent-project.github.io/), except for one important part: in `reagent`, multiple components can share one `atom` so that all of them re-render when it changes, whereas `useState` only creates state local to the component that called it and only that component (and its children) re-render when that state changes. To say it another way, React hooks offer little to no affordance for doing scalable global state management in the fashion of `redux` or `mobx`, whereas `reagent` `atom`s do. However, I think it would be relatively simple to implement `atom`s for React with a light abstraction on top of `useState` and `useMutationEffect`.

I've already implemented a minimal version of `Atom` for `react`, and so far I really like the result. I'm curious to know what others think.  

I'm just calling it **[react-atom](https://github.com/derrickbeining/react-atom)** for now and **[have published it (with TypeScript typings) to `npm`](https://www.npmjs.com/package/@dbeining/react-atom)**. Here are **[the docs](https://derrickbeining.github.io/react-atom/)**. And here is a **[JS CodeSandbox](https://codesandbox.io/s/m3x9wn6kmy)** and **[TS CodeSandbox](https://codesandbox.io/s/km72yynqov)** to try it out live. If you're wondering, the API for it is mostly just plagiarized from `clojure/script` atoms.

I think there would be a lot of benefit to `react.js` providing something like this as an official way for doing global state management. Doing so would really help unify the community and library ecosystem, and it would allow some performance optimizations to be managed internally by `react`. It could be made more plug-able too, so as to enable the various niceties made possible by `redux` middleware.

The new Hooks proposal seems like the perfect opportunity to introduce something like this. What do you think?

**Comment by derrickbeining** [comment]
Here is a basic example of what using an `Atom` looks like (remember, it's just using `useState` and `useMutationEffect` internally):

```jsx
import React from 'react';
import ReactDOM from 'react-dom';
import {Atom, deref, swap} from '@dbeining/react-atom';

/////////////////////// APP STATE /////////////////////////
const stateAtom = Atom.of({
  count: 0,
  text: '',
  data: {
    // ...just imagine
  },
});

/////////////////////// EFFECTS /////////////////////
/**
 * `swap` applies a function to an Atom's current state to set it's next state 
 * Then it rerenders all the function components that `deref`ed that atom
 * so they read its new state
 */
const increment = () =>
  swap(stateAtom, (state) => ({...state, count: state.count + 1}));

const decrement = () =>
  swap(stateAtom, (state) => ({...state, count: state.count - 1}));

const updateText = (evt) =>
  swap(stateAtom, (state) => ({...state, text: evt.target.value}));

const loadSomething = () =>
  fetch('https://jsonplaceholder.typicode.com/todos/1')
    .then((res) => res.json())
    .then((data) => swap(stateAtom, (state) => ({...state, data})))
    .catch(console.error);

///////////////// COMPONENT  /////////////////////
export const App = () => {
  /**
   * `deref` reads the current value of the atom
   * also, on the first render, it subscribes the component to the Atom
   * so that it re-renders when that Atom changes. 
   * It will automatically unsubscribe from updates when the component un-mounts.
   */
  const {count, data, text} = deref(stateAtom);

  return (
    <div>
      <h2>Count: {count}</h2>
      <h2>Text: {text}</h2>

      <button onClick={increment}>Moar</button>
      <button onClick={decrement}>Less</button>
      <button onClick={loadSomething}>Load Data</button>
      <input type="text" onChange={updateText} value={text} />
      <p>{JSON.stringify(data, null, "  ")}</p>
    </div>
  );
};

ReactDOM.render(<App />, document.getElementById('root'));
```

**Comment by mfrachet** [comment]
Doesn't this proposal provide a solution? https://github.com/reactjs/rfcs/pull/89

**Comment by derrickbeining** [comment]
@mfrachet , yeah that's what @gaearon pointed out to me. #89 wasn't posted until the day after I finished prototyping `react-atom` so I missed it when I posted here.  `Context.write` does seem to be getting at a similar problem, but I'm curious what differences there are in the implications for each approach.

`react-atom` simply uses `useState` hooks internally to trigger component re-renders on state change, so it works with concurrent rendering and Suspense to the same degree that `useState` does, and as far as I can tell,`react-atom` has no problem sharing/syncing state across multiple React roots -however, I probably don't understand well enough the issues @acdlite described about concurrent rendering, Suspense, and dealing with multiple roots.

@acdlite , if you happen to see this, I'd be super grateful to know what you think about `react-atom` and its viability as an alternative to enhancing the `Context` API.
